### PR TITLE
feature: add doc, remove reverse order

### PR DESCRIPTION
**Pull request by fwx5618177**
- Add a doc to give an introduction about this project.
- Fix error about struct in file.zig.
- Remove reverse order.

### CODE REVIEW

**Comment by fwx5618177** [comment]
@Hejsil  PTAL.

File: ReadMe_cn.md [hunk 1]
```diff
@@ -0,0 +1,985 @@
+# 【Zig】乐器数字接口-Midi
```
Comment (Hejsil) [inline ReadMe_cn.md:1]: While I respect the effort in writing these docs, I cannot accept a readme in a language I don't understand. It is not really something I can maintain (even though I don't really work on this anymore) and it could anything. I could run it through some automated translation to get an idea of the meaning but that is still quite imperfect.

File: midi.zig [hunk 2]
```diff
@@ -2,25 +2,27 @@ const std = @import("std");
...
-
 pub const Message = struct {
     status: u7,
     values: [2]u7,
 
     pub fn kind(message: Message) Kind {
-        const _kind = @truncate(u3, message.status >> 4);
-        const _channel = @truncate(u4, message.status);
+        const _kind = @as(u3, @truncate(message.status >> 4));
+        const _channel = @as(u4, @truncate(message.status));
```
Comment (Hejsil) [inline midi.zig:25]: These are all over. I recommend specifying the type of the variable instead of using `@as`

```suggestion
        const _kind: u3 = @truncate(message.status >> 4);
        const _channel: u4 = @truncate(message.status);
```

File: midi/file.zig [hunk 3]
```diff
@@ -112,8 +114,9 @@ pub const TrackIterator = struct {
         it.last_event = event;
 
         const start = it.stream.pos;
-        switch (event.kind) {
-            .MetaEvent => |meta_event| {
+
+        var end: usize = switch (event.kind) {
```
Comment (Hejsil) [inline midi/file.zig:118]:
```suggestion
        const end = switch (event.kind) {
```

**Review by Hejsil — COMMENTED** [review]
Looks good. Do note this is not really something I'm working on anymore. I'm still happy to merge improvements though.

**Comment by Hejsil** [comment]
I also update cicd to use zig master instead of the old `0.8.0`. I recommend rebasing so cicd can try again on this branch :)

File: ReadMe_cn.md (hunk 1 above)
Comment (fwx5618177) [reply ReadMe_cn.md:1]: > While I respect the effort in writing these docs, I cannot accept a readme in a language I don't understand. It is not really something I can maintain (even though I don't really work on this anymore) and it could contain anything. I could run it through some automated translation to get an idea of the meaning but that is still quite imperfect.

I would rewrite it in English. Don't worry.

File: midi.zig [hunk 4]
```diff
@@ -50,7 +52,7 @@ pub const Message = struct {
 
     pub fn channel(message: Message) ?u4 {
         const _kind = message.kind();
-        const _channel = @truncate(u4, message.status);
+        const _channel = @as(u4, @truncate(message.status));
```
Comment (Hejsil) [inline midi.zig:55]: Aah sorry, I was trying to imply in a previous comment that there are a lot of these places where this needs fixing. I'll just point them out :)
```suggestion
        const _channel: u4 = @truncate(message.status);
```

File: midi.zig [hunk 5]
```diff
@@ -88,8 +90,8 @@ pub const Message = struct {
 
     pub fn setValue(message: *Message, v: u14) void {
         message.values = .{
-            @truncate(u7, v >> 7),
-            @truncate(u7, v),
+            @as(u7, @truncate(v >> 7)),
+            @as(u7, @truncate(v)),
```
Comment (Hejsil) [inline midi.zig:94]:
```suggestion
            @truncate(v >> 7),
            @truncate(v),
```

File: midi/decode.zig [hunk 6]
```diff
@@ -31,8 +31,8 @@ pub fn message(reader: anytype, last_message: ?midi.Message) !midi.Message {
         break :blk m.status;
     } else return error.InvalidMessage;
 
-    const kind = @truncate(u3, status_byte >> 4);
-    const channel = @truncate(u4, status_byte);
+    const kind = @as(u3, @truncate(status_byte >> 4));
+    const channel = @as(u4, @truncate(status_byte));
```
Comment (Hejsil) [inline midi/decode.zig:35]:
```suggestion
    const kind: u3 = @truncate(status_byte >> 4);
    const channel: u4 = @truncate(status_byte);
```

File: midi/decode.zig [hunk 7]
```diff
@@ -118,8 +118,8 @@ pub fn int(reader: anytype) !u28 {
     var res: u28 = 0;
     while (true) {
         const b = try reader.readByte();
-        const is_last = @truncate(u1, b >> 7) == 0;
-        const value = @truncate(u7, b);
+        const is_last = @as(u1, @truncate(b >> 7)) == 0;
+        const value = @as(u7, @truncate(b));
```
Comment (Hejsil) [inline midi/decode.zig:122]:
```suggestion
        const value: u7 = @truncate(b);
```

File: midi/encode.zig [hunk 8]
```diff
@@ -102,18 +109,18 @@ pub fn file(writer: anytype, f: midi.File) !void {
     try writer.writeAll(&encode.fileHeaderToBytes(.{
         .chunk = .{
             .kind = midi.file.Chunk.file_header.*,
-            .len = @intCast(u32, midi.file.Header.size + f.header_data.len),
+            .len = @as(u32, @intCast(midi.file.Header.size + f.header_data.len)),
```
Comment (Hejsil) [inline midi/encode.zig:112]:
```suggestion
            .len = @intCast(midi.file.Header.size + f.header_data.len),
```

File: midi/encode.zig [hunk 9]
```diff
@@ -102,18 +109,18 @@ pub fn file(writer: anytype, f: midi.File) !void {
...
         },
         .format = f.format,
-        .tracks = @intCast(u16, f.chunks.len),
+        .tracks = @as(u16, @intCast(f.chunks.len)),
```
Comment (Hejsil) [inline midi/encode.zig:115]:
```suggestion
        .tracks = @intCast(f.chunks.len),
```

File: midi/encode.zig [hunk 10]
```diff
@@ -102,18 +109,18 @@ pub fn file(writer: anytype, f: midi.File) !void {
...
         .division = f.division,
     }));
     try writer.writeAll(f.header_data);
 
     for (f.chunks) |c| {
         try writer.writeAll(&encode.chunkToBytes(.{
             .kind = c.kind,
-            .len = @intCast(u32, c.bytes.len),
+            .len = @as(u32, @intCast(c.bytes.len)),
```
Comment (Hejsil) [inline midi/encode.zig:123]:
```suggestion
            .len = @intCast(c.bytes.len),
```

File: midi/encode.zig [hunk 11]
```diff
@@ -63,18 +63,25 @@ pub fn fileHeaderToBytes(header: midi.file.Header) [14]u8 {
...
-    while (tmp != 0 or is_first) : (is_first = false) {
-        fbs.writeByte(@truncate(u7, tmp) | (@as(u8, 1 << 7) * @boolToInt(!is_first))) catch
-            unreachable;
-        tmp >>= 7;
+
+    while (byteCount > 0) : (byteCount -= 1) {
+        const shifted = tmp >> (7 * (byteCount - 1));
+
+        if (is_first) {
+            try writer.writeByte(@as(u8, @truncate(shifted)));
```
Comment (Hejsil) [inline midi/encode.zig:79]:
```suggestion
            try writer.writeByte(@truncate(shifted));
```

File: ReadMe.md [hunk 12]
```diff
@@ -0,0 +1,236 @@
+# Zig-midi
```
Comment (Hejsil) [inline ReadMe.md:1]: Thanks for the translation, now I can actually review the content.

I don't think a description of the details of the midi format belongs in a readme for a library encoding and decoding such format. A link to external resources would be fine though. The sections `Basic`, `MIDI Message Structure` are heavy on this.

A lot of the other information would be better served as doc comments in the code itself. For example the section describing `Message.Kind` enumerations should be doc comments in the code instead. I think most other sections fall into this category.

To summarize, this is not really the way I would prefer a readme to be structured. Something along the lines of https://github.com/Hejsil/zig-clap is my the style I prefer, but I have to admit that readme has also become quite verbose.

**Comment by fwx5618177** [comment]
> Here is an idea. How about we get the improvements merged first. Then you can open a new PR for documentation improvements, so that my nitpicks on that wont block the merging of the rest.

I think it's awesome. We could merge it firstly, then I would modify the content of this readme.

**Comment by Hejsil** [comment]
Merge [here](https://github.com/Hejsil/zig-midi/commit/88e2be5569f8a59b23f118f8d77db3919302b5e2) without the readme and + some extra changes to make this compile with zig master.

Had to revert changes `midi.encode.int` as they where failing tests.
